### PR TITLE
Behavior tree application with reactive GUI

### PR DESCRIPTION
**Pull request by marip8**
This PR replaces the custom application back-end logic with a behavior tree to allow the workflow to be more modular and customizable. This PR also updates the GUI such that the behavior tree can change its appearance based on the current behavior that is being executed

https://github.com/ros-industrial-consortium/scan_n_plan_workshop/assets/18411310/9ad678e1-d265-4d7e-8420-aa8b11f00609

<details>
  <summary>Main BT</summary>
  
  image

</details>
<details>
  <summary>Process BT</summary>

  image

</details>
<details>
  <summary>Motion BT</summary>

  image

</details>
<details>
  <summary>Scan Motion BT</summary>

  image

</details>

### CODE REVIEW

**Comment by marip8** [comment]
@marrts I tested in `humble` and everything seems to work appropriately

**Comment by marip8** [comment]
I also noticed that #74 is no longer an issue in `humble` with these changes

**Comment by gavanderhoorn** [comment]
Very nice work :+1:

File: snp_blending/config/controllers.yaml [hunk 1]
```diff
@@ -24,8 +24,8 @@ joint_trajectory_position_controller:
 
     command_interfaces:
       - position
-      - velocity
-      - acceleration
+#      - velocity
+#      - acceleration
```
Comment (marrts) [inline snp_blending/config/controllers.yaml:28]: If these interfaces aren't available we can probably delete

File: snp_automate_2022/launch/start.launch.xml [hunk 2]
```diff
@@ -74,9 +84,9 @@
   </group>
 
   <!-- Launch Toolpath Planning Node -->
-  <node pkg="snp_tpp" exec="snp_tpp_app">
+<!--  <node pkg="snp_tpp" exec="snp_tpp_app">
     <param name="config_file" value="$(find-pkg-share snp_automate_2022)/config/tpp.yaml"/>
-  </node>
+  </node>-->
```
Comment (marrts) [inline snp_automate_2022/launch/start.launch.xml:89]: Should remove this now that it is inside the GUI inherently

File: snp_blending/launch/start.launch.xml [hunk 3]
```diff
@@ -75,9 +86,9 @@
   </group>
 
   <!-- Launch Toolpath Planning Node -->
-  <node pkg="snp_tpp" exec="snp_tpp_app">
+<!--  <node pkg="snp_tpp" exec="snp_tpp_app">
     <param name="config_file" value="$(find-pkg-share snp_blending)/config/tpp.yaml"/>
-  </node>
+  </node>-->
```
Comment (marrts) [inline snp_blending/launch/start.launch.xml:91]: Remove

File: snp_blending/launch/start.launch.xml [hunk 4]
```diff
@@ -88,9 +99,9 @@
   </include>
 
   <!-- Launch Motion Execution Node -->
-  <node pkg="snp_motion_execution" exec="motion_execution_server">
+<!--  <node pkg="snp_motion_execution" exec="motion_execution_server">
     <param name="follow_joint_trajectory_action" value="$(var follow_joint_trajectory_action)"/>
-  </node>
+  </node>-->
```
Comment (marrts) [inline snp_blending/launch/start.launch.xml:104]: Is this now a BT in the GUI?

Comment (marrts) [reply snp_blending/launch/start.launch.xml:104]: I see it's still getting launched in the snp_automate_2022 `start.launch`

File: snp_scanning/scripts/scan_motion_plan_from_file_node [hunk 5]
```diff
@@ -20,18 +36,31 @@ class ScanMotionPlanFromFileNode(Node):
...
+            header.frame_id = scan_traj['header']['frame_id']
+
+            # Approach
+            response.approach.header = header;
+            response.approach.joint_names = scan_traj['joint_names']
+            for i in range(0, 2):
+              response.approach.points.append(create_point(points[i]))
+
+            # Process
+            # TODO: adjust time
```
Comment (marrts) [inline snp_scanning/scripts/scan_motion_plan_from_file_node:52]: What does "adjust time" mean in this context?

File: snp_application/include/snp_widget.h [hunk 6]
```diff
@@ -1,107 +1,37 @@
...
-signals:
-  void log(const QString& message);
-  void updateStatus(bool success, QString current_process, QString next_process, unsigned step);
+  rclcpp::Node::SharedPtr bt_node_;
+  rclcpp::Node::SharedPtr tpp_node_;
+  rclcpp::executors::SingleThreadedExecutor tpp_node_executor_;
+  std::future<void> tpp_node_future_;
+  Ui::SNPWidget* ui_;
+  BT::Blackboard::Ptr board_;
+  BT::BehaviorTreeFactory factory_;
```
Comment (marrts) [inline snp_application/include/snp_widget.h:33]: Maybe change name to `bt_factory_`

File: snp_application/src/snp_widget.cpp [hunk 7]
```diff
@@ -1,686 +1,197 @@
...
-
-void SNPWidget::onScanApproachDone(FJTResult result)
-{
-  snp_msgs::srv::ExecuteMotionPlan::Response::SharedPtr response = result.get();
-  if (response->success)
-  {
-    emit updateStatus(true, SCAN_APPROACH_ST, START_RECONSTRUCTION_ST, STATES.at(START_RECONSTRUCTION_ST));
-    emit log("Successfully executed scan approach motion");
+    auto* vl = new QVBoxLayout(this);
+    vl->addWidget(w);
```
Comment (marrts) [inline snp_application/src/snp_widget.cpp:42]: This might've come from an example somewhere, but variable names `w` and `vl` aren't really descriptive.

**Review by marrts — COMMENTED** [review]
Haven't had the chance to run it locally, but I really like the way this is looking

File: snp_blending/launch/start.launch.xml (hunk 4 above)
Comment (marip8) [reply snp_blending/launch/start.launch.xml:104]: It should be removed in both. The BT now does all of the steps that this service node was doing. It probably makes sense to just get rid of this node now

File: snp_scanning/scripts/scan_motion_plan_from_file_node (hunk 5 above)
Comment (marip8) [reply snp_scanning/scripts/scan_motion_plan_from_file_node:52]: It's already accounted for in `time_offset` [here](https://github.com/marip8/scan_n_plan_workshop/blob/cf60b557fa079c01d16ed607cf46b3bd997c7fc3/snp_scanning/scripts/scan_motion_plan_from_file_node#L13). I'll remove the comments

File: snp_application/src/snp_widget.cpp [hunk 8]
```diff
@@ -1,686 +1,201 @@
...
-  {
-    geometry_msgs::msg::PoseArray flat_toolpath_msg;
-    flat_toolpath_msg.header.frame_id = reference_frame_;
-    for (auto& toolpath : tool_paths_)
-    {
-      for (auto& segment : toolpath.segments)
+    auto* thread = new BTThread(this);
+
+    // Create the BT factory
+    BT::BehaviorTreeFactory factory;
```
Comment (marrts) [inline snp_application/src/snp_widget.cpp:134]: Can you change this to `bt_factory`

**Review by marrts — COMMENTED** [review]
Ran it locally now and everything works great, love the new interface.

**Comment by marip8** [comment]
Strange issues from `rolling` build that seem to be related to `tesseract_ros2`. Interesting that the `rolling` Docker job succeeds though. I'll log this and look into it with a separate PR

> 2024-02-08T17:56:28.8943289Z /usr/bin/ld: /opt/tesseract_ros2/install/tesseract_monitoring/lib/libtesseract_monitoring_environment.so: undefined reference to `rclcpp::experimental::IntraProcessManager::add_publisher(std::shared_ptr<rclcpp::PublisherBase>)'

>2024-02-08T17:56:28.8945198Z /usr/bin/ld: /opt/tesseract_ros2/install/tesseract_monitoring/lib/libtesseract_monitoring_environment.so: undefined reference to `rclcpp::GuardCondition::GuardCondition(std::shared_ptr<rclcpp::Context>, rcl_guard_condition_options_s)'

>2024-02-08T17:56:28.8947461Z /usr/bin/ld: /opt/tesseract_ros2/install/tesseract_monitoring/lib/libtesseract_monitoring_environment.so: undefined reference to `rclcpp::experimental::IntraProcessManager::add_subscription(std::shared_ptr<rclcpp::experimental::SubscriptionIntraProcessBase>)'

**Pull request merged**